### ISSUE TITLE
Automated transformations break log_joint computation

### ISSUE THREAD

**Issue by davmre** [blei-lab/edward#807]
Currently all inference in models with auto-transformed support is silently producing meaningless results. This seems quite serious because automated transformations are enabled by default. 

The problem: when an automated transformation is applied to a latent variable `z`, inference methods that use the `ed.copy(x, dict_swap)` idiom to compute a log joint density (this is basically all inference methods as far as I can tell) will not swap out the transformed variable correctly. Any other variable `x` that depends on `z` will be evaluated using its original dependency---a prior sample from `z`---instead of at a posterior value `qz`. This means that instead of being a deterministic quantity that depends on the posterior, the constructed `log_joint` becomes a random quantity that has no relationship to `qz` outside of the explicit prior term (that term `z.log_prob(qz)` will be computed correctly since it accesses `qz` explicitly rather than through `dict_swap`). This is not caught by tests because none of the tests for automated transformations consider models with multiple variables. 

More specifically, the `Inference` class replaces `z` with `transform(z)` in its `latent_vars` mapping. This is inherited by `dict_swap` which also then contains `transform(z)` rather than `z`. And since `z` isn't in `dict_swap`, its occurences in the graph will never be replaced by `qz`.

I've reproduced this problem with HMC and KLqp inference but it should exist more generally. Here's a test case running HMC on a simple beta-bernoulli model:
```python
import numpy as np
import tensorflow as tf
import edward as ed

# model
z = ed.models.Beta(1., 1., name="z")
xs = ed.models.Bernoulli(probs=z, sample_shape=10)

# data
x_obs = np.asarray([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=np.int32)

# inference
qz_samples = tf.Variable(tf.random_uniform(shape=(2000,)))
qz = ed.models.Empirical(params=qz_samples, name="z_posterior")
inference_hmc = ed.inferences.HMC({z: qz}, data={xs: x_obs})
inference_hmc.run(step_size=1.0, n_steps=5, auto_transform=True)

z_unconstrained = inference_hmc.transformations[z]
qz_constrained = z_unconstrained.bijector.inverse(qz_samples)
qz_mean, qz_var = sess.run(tf.nn.moments(qz_constrained, 0))

true_posterior = ed.models.Beta(1. + np.sum(x_obs), 1. + np.sum(1-x_obs))
pz_mean, pz_var = sess.run((true_posterior.mean(), true_posterior.variance()))

print("true mean {} vs inferred {}".format(pz_mean, qz_mean))
print("true var {} vs inferred {}".format(pz_var, qz_var))
```

I'm working on a PR that should fix this issue by ensuring that `inference.latent_vars` always maps the actual model variables `z` to valid `qz` distributions having the same support as the model variables, and adds another dict `inference.latent_vars_unconstrained` to track the counterparts of these distributions with unconstrained support. This seemed like the easiest path that doesn't require modifying log_joint calculations in every inference method. Thoughts welcome!

**Comment by davmre** [comment]
Update: see PR https://github.com/blei-lab/edward/pull/808

**Comment by dustinvtran** [comment]
Fixed in #808.
